#### Exercise 4.7 Deriving new variables

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
#Creating string of project folder.

path = r'/Users/nautis1/data_project/09-2025-Instacart-Basket-Analysis/Instacart Basket Analysis'

In [3]:
#import the merged data frame
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 data' , 'prepared data' , 'ords_prods_merge.pkl'))

In [4]:
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   3139998      138    prior            28          6                 11   
1   1977647      138    prior            30          6                 17   
2    389851      709    prior             2          0                 21   
3    652770      764    prior             1          3                 13   
4   1813452      764    prior             3          4                 17   

   days_since_prior_order  customer_first_order  add_to_cart_order  reordered  \
0                     3.0                 False                  5          0   
1                    20.0                 False                  1          1   
2                     6.0                 False                 20          0   
3                     NaN                  True                 10          0   
4                     9.0                 False                 11          1   

  _merge  
0   both  
1   both  
2   both  
3   both  
4   both

In [5]:
#create a price label column based on different conditions
ords_prods_merge.loc[ords_prods_merge['prices'] >15, "price_label"] = "High-range product"

In [6]:
ords_prods_merge.loc[(ords_prods_merge['prices'] <=15) & (ords_prods_merge['prices'] > 5), 'price_label'] = 'Mid-range product'

In [7]:
ords_prods_merge.loc[ords_prods_merge['prices'] <= 5, 'price_label'] = 'Low-range product'

In [8]:
ords_prods_merge['price_label'].value_counts()

price_label
Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

In [9]:
#loop through to create busiest day labels 
result = []

for value in ords_prods_merge['order_dow']:
    if value == 0:
        result.append("Busiest day")
    elif value == 4:
        result.append("Least busy")
    else:
        result.append("Regularly busy")

In [10]:
#create a column to save the result
ords_prods_merge['busiest_day'] = result

In [11]:
ords_prods_merge['busiest_day'].value_counts()

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

In [12]:
ords_prods_merge['order_dow'].value_counts(dropna = False)

order_dow
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

#### Task 2

In [13]:
#loop through to create new busiest day labels 
result = []

for value in ords_prods_merge['order_dow']:
    if value == 0 or value == 1:
        result.append("Busiest days")
    elif value == 4 or value == 3:
        result.append("Least busy")
    else:
        result.append("Regularly busy")

In [14]:
#create a column to save the new results
ords_prods_merge['busiest_days'] = result

In [15]:
ords_prods_merge['busiest_days'].value_counts()

busiest_days
Regularly busy    12916111
Busiest days      11864412
Least busy         7624336
Name: count, dtype: int64

The values for each of the busy day category is changed. The sum of 0 and 1 day of the week makes up the busiet days value 11864412. Similar is the true for lease busy day as well. 

#### Task 5

When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

In [16]:
ords_prods_merge['order_hour_of_day'].value_counts()

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

In [17]:
#loop through to create new busiest hour labels 
result = []

for value in ords_prods_merge['order_hour_of_day']:
    if (value >= 9) & (value <= 17):
        result.append("Most orders")
    elif (value >= 2) & (value <= 5):
        result.append("Fewest orders")
    else:
        result.append("Average orders")

In [18]:
#create the new column
ords_prods_merge['busiest_period_of_day'] = result

In [19]:
#freq of new column
ords_prods_merge['busiest_period_of_day'].value_counts()

busiest_period_of_day
Most orders       23205725
Average orders     8937275
Fewest orders       261859
Name: count, dtype: int64

In [20]:
# Export data to pkl
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_new_var.pkl'))